### DISCUSSION TITLE
Force visit both nodes or none at all

### DISCUSSION THREAD

**Discussion by glebarez** [google/or-tools#3366] [Routing (and legacy CP) questions] [answered]
I'm having trouble implementing the following constraint:
- node B must be visited only if node A was visited (if A is dropped from solution, the B must also be dropped).

sounds simple, but when I add the following:
```python
solver.Add(rm.ActiveVar(a_idx) == rm.ActiveVar(b_idx))
```
the solver struggles to move beyond initial AllUnperformed solution.

I also tried many variations like
```python
activeCount = rm.ActiveVar(a_idx) + rm.ActiveVar(b_idx)
bothOrNone = solver.IsMemberVar(activeCount, [0, 2])
self.solver.Add(bothOrNone == 1)
```
with no success.

It seems that constraint like this just blocks solver from finding solutions (at least in adequate times).
Any ideas for performance-efficient implementation of such constraint?

**Comment by Mizux** [comment]
question: How did you allow the dropping of these nodes ?
did you use `max_cardinality=2` in `routing.AddDisjunction()` ?

**Reply by glebarez** [reply]
I just use ```AddDisjunction([idx], penalty)``` basically for every node in the model.

**Comment by jmarca** [comment]
First double check that it is possible to visit both nodes.

Then maybe try a conditional constraint, where the requirement that B is active only holds if A is active, and also if B then A

You also might try using the AddPickupAndDeliver API call to let the solver know these are paired. Because pickup and delivery is exactly that—both on or both off—and the API call tells the solver to add both to a solution (or remove both) at once. 

James

**Reply by glebarez** [reply]
Hello, @jmarca.
Yeah, the model is totally solvable, given enough time it copes without dropping any nodes.
But when time is short, it finishes with some nodes dropped, which is okay, but I need to make it drop B, if A is dropped.

I tried AddPickupAndDeliver for this two nodes, but it causes the solver to just hang (?!). It stops logging solutions and hangs way after the time limit.
**BTW this nodes are visited by different vehicles** (which is naturally implemented as with demands of different dimensions).
I

I also tried the following conditional with no success:
```python
solver.Add(
                solver.ConditionalExpression(
                    rm.ActiveVar(A_idx) == 0,
                    rm.ActiveVar(B_idx) == 0,
                    1,
                )
                == 1
            )
```

PS: I had the same issue when I was trying to play with constraint of kind ```NextVar(x) == y```.
It seems to me that there's some unwritten rule that sound like ```don't mess with NextVar / ActiveVar constraints```.
@Mizux any insights on that part?

**Reply by glebarez** [reply]
@jmarca @Mizux I tested in with various sizes of models.
Apparently ```solver.Add(rm.ActiveVar(a_idx) == rm.ActiveVar(b_idx))``` works as expected, in small models, and does exactly what I demand. But in a bigger model, even the one that previously was solved under 10 seconds, adding such a constraint just makes it time-infeasible.
Any guidance on a probable cause of such performance drop?

**Comment by jmarca** [comment, marked as answer]
I missed that it can be different vehicles.

I *think* the issue is that the solver gets stuck trying to rotate nodes in and out.  

Suppose A and B and both not visited.

If it adds node A for vehicle i, then it has to somehow also add node B to some other vehicle in the same candidate solution/improvement.

This is probably not possible every time.

What I would do is hack around to try to figure out a way that the solver can add A and not add B without getting into a forbidden state, but then make it really really beneficial for the solver to add B to the solution somewhere if A is added.  

Maybe something like a disjunction with A, A' and B with cardinality of 2, where A' is always reachable from A, but is expensive in some way, so the solver will try hard to reach B if it has reached A.  And vice versa.

**Reply by glebarez** [reply]
Thank you @jmarca  for the answer.
Indeed the solver seems to be stuck trying to rotate the nodes, if we use simplest ```solver.Add(rm.ActiveVar(a_idx) == rm.ActiveVar(b_idx))``` constraint. 
Currently the variant with 
```python
solver.Add(
                solver.ConditionalExpression(
                    rm.ActiveVar(A_idx) == 0,
                    rm.ActiveVar(B_idx) == 0,
                    1,
                )
                == 1
            )
``` 
works for me, though it's only a half effect of what I demand (Adding the reversed condition also makes solver stuck)

PS: overall lesson: use constraints for Active / Vehicle vars carefully.
Another example of what I observed:
Just adding simple constraint with ```AddAllowedAssignments``` pushes solver to fall for suboptimal solutions.
I think another way would be to just write a custom DecisionBuilder for some cases like this, but Python API doesn't expose much of that.

**Comment by jmarca** [comment]
I know it is the same thing, but I would try the positive variant:

Condition Active(Aidx) == 1 then Active(b) == 1

And

Condition Active(Bidx) == 1 then Active(Aidx) == 1


But really try the pickup and delivery constraint call.  https://github.com/google/or-tools/blob/6897c64dbf7d18b97d3563d28dc4820e3eb52022/ortools/constraint_solver/routing.cc#L2028

It does stuff under the hood to consider pairs of nodes as both active or both inactive, which is what you want. Without that, I think the solver does not know to rotate in/out both locations at the same time, resulting in impossible states.

If it only adds one node at a time, it cannot get to a state where both nodes are active from a state where both nodes are inactive. 

This is a common irritant for me. 

James
